### PR TITLE
DatePicker, LoopingSelector & SnapPoints

### PR DESCRIPTION
**Pull request by carldebilly**
Implements #1390
Related to #365

# Features in this PR
- `<DatePicker />`: Previously on Uno, only native pickers were available, so nothing on WASM.
- `LoopingSelector`: used by some internal controls, like the `DatePicker`
- _Snap Points_: Support for `Mandatory` snappoints on `ScrollViewer`

Previous native pickers are still available (using the same approach as other native controls), but the default one is the new "WinUI" one. This is a direct port of the WinUI C++ original code.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [X] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [X] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [X] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

### CODE REVIEW

File: src/SamplesApp/SamplesApp.UITests/SamplesApp.UITests.csproj [hunk 1]
```diff
@@ -5,11 +5,10 @@
...
 
 	<ItemGroup> 
-		<PackageReference Include="FluentAssertions" Version="5.10.3" />
-		<PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.8.3" />
+		<PackageReference Include="FluentAssertions" />
+		<PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.8.0" />
 		<PackageReference Include="NUnit" />
 		<PackageReference Include="NUnit3TestAdapter" /> 
-		<PackageReference Include="Xamarin.UITest" Version="3.0.12" />
 		<PackageReference Include="Uno.UITest" />
```
Comment (MartinZikmund) [inline src/SamplesApp/SamplesApp.UITests/SamplesApp.UITests.csproj:12]: Are these changes intentional?

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/DatePicker/DatePickerSample.xaml [hunk 2]
```diff
@@ -14,7 +18,8 @@
 		<DatePicker Grid.Row="1"
 		            Margin="15,10"
 		            HorizontalAlignment="Center"
-		            VerticalAlignment="Top">
+		            VerticalAlignment="Top"
+					Date="{Binding Date}">
 		</DatePicker>
 
 		<!-- DatePicker is broken: https://github.com/unoplatform/uno/issues/188 -->
```
Comment (MartinZikmund) [inline src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/DatePicker/DatePickerSample.xaml:25]: This is probably no longer applicable

File: src/Uno.UI.RuntimeTests/MUX/Helpers/ControlHelper.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+				;
+
+				//if (button == null)
+				//{
+				//	var cmdBar = TestServices.Utilities.RetrieveOpenBottomCommandBar();
+				//	if (cmdBar != null)
+				//	{
+				//		button = cmdBar.PrimaryCommands[isAccept ? 0 : 1] as ButtonBase;
+				//	}
+				//}
```
Comment (MartinZikmund) [inline src/Uno.UI.RuntimeTests/MUX/Helpers/ControlHelper.cs:82]: This can be removed

File: src/Uno.UI.Tests/Windows_UI_XAML_Controls/DatePickerTests/Given_DatePicker.cs [hunk 4]
```diff
@@ -13,13 +13,5 @@ namespace Uno.UI.Tests.DatePickerTests
 	[TestClass]
 	public class Given_DatePicker
 	{
```
Comment (MartinZikmund) [inline src/Uno.UI.Tests/Windows_UI_XAML_Controls/DatePickerTests/Given_DatePicker.cs:15]: Seems like the file can go away

File: src/Uno.UI/UI/Xaml/Controls/CommandBar/CommandBarFlyout.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+
+namespace Windows.UI.Xaml.Controls
+{
+	partial class CommandBarFlyout
+	{
+		protected override Control CreatePresenter() => throw new NotImplementedException();
```
Comment (MartinZikmund) [inline src/Uno.UI/UI/Xaml/Controls/CommandBar/CommandBarFlyout.cs:7]: Add [Notimplemented] attribute

File: src/Uno.UI/UI/Xaml/Controls/Control/Control.cs [hunk 6]
```diff
@@ -1117,5 +1117,28 @@ private static bool GetIsEventOverrideImplemented(Type type, string name, Type[]
 
 		public string[] CurrentVisualStates => VisualStateGroups.Select(vsg => vsg.CurrentState?.Name).ToArray();
 #endif
+
+		internal void ConditionallyGetTemplatePartAndUpdateVisibility<T>(
+			string strName,
+			bool visible,
+			ref T element) where T:UIElement
+        {
```
Comment (MartinZikmund) [inline src/Uno.UI/UI/Xaml/Controls/Control/Control.cs:1125]: Indentation is somehow weird here

File: src/Uno.UI/UI/Xaml/Controls/DatePicker/DatePicker.Properties.cs [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+					propertyChangedCallback: (s, e) => ((DatePicker)s).OnDatePropertyChanged((DateTimeOffset)e.NewValue, (DateTimeOffset)e.OldValue)));
+
+		private void OnDatePropertyChanged(DateTimeOffset newValue, DateTimeOffset oldValue)
+		{
+//#if XAMARIN
+//			UpdateDisplayedDate();
+//#endif
+
+			OnDateChangedPartial();
+			//DateChanged?.Invoke(this, new DatePickerValueChangedEventArgs(newValue, oldValue));
```
Comment (MartinZikmund) [inline src/Uno.UI/UI/Xaml/Controls/DatePicker/DatePicker.Properties.cs:34]: Remove commented out code

File: src/Uno.UI/UI/Xaml/Controls/WinUIDatePicker/DatePicker.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
+using System;
```
Comment (MartinZikmund) [inline src/Uno.UI/UI/Xaml/Controls/WinUIDatePicker/DatePicker.cs:1]: This file can probably be removed

File: src/Uno.UI/Uno.UI.csproj [hunk 9]
```diff
@@ -184,9 +184,15 @@
...
 	<ItemGroup>
+	  <None Update="C:\src\Uno2\src\Uno.UI\UI\Xaml\Controls\DatePicker\DatePicker_themeresources.xaml">
+	    <Generator>MSBuild:Compile</Generator>
+	  </None>
 	  <None Update="Mixins\Wasm\VisualStatesMixins.tt">
 	    <LastGenOutput>VisualStatesMixins.g.cs</LastGenOutput>
 	  </None>
+	  <None Update="UI\Xaml\Controls\DatePicker\DateTimePickerFlyout_themeresources.xaml">
+	    <Generator>MSBuild:Compile</Generator>
+	  </None>
```
Comment (MartinZikmund) [inline src/Uno.UI/Uno.UI.csproj:195]: Changes here can be reverted

File: src/Uno.UWP/Uno.csproj [hunk 10]
```diff
@@ -42,9 +42,21 @@
...
+	  <UpToDateCheckInput Remove="Globalization\DateTimeFormatting\DayFormat.cs" />
+	  <UpToDateCheckInput Remove="Globalization\DateTimeFormatting\DayOfWeekFormat.cs" />
+	  <UpToDateCheckInput Remove="Globalization\DateTimeFormatting\HourFormat.cs" />
+	  <UpToDateCheckInput Remove="Globalization\DateTimeFormatting\MinuteFormat.cs" />
+	  <UpToDateCheckInput Remove="Globalization\DateTimeFormatting\MonthFormat.cs" />
+	  <UpToDateCheckInput Remove="Globalization\DateTimeFormatting\SecondFormat.cs" />
+	  <UpToDateCheckInput Remove="Globalization\DateTimeFormatting\YearFormat.cs" />
+	</ItemGroup>
+
 	<ItemGroup>
```
Comment (MartinZikmund) [inline src/Uno.UWP/Uno.csproj:60]: Revert these

**Review by MartinZikmund — APPROVED** [review]
ClapApplauseGIF

File: src/SamplesApp/SamplesApp.UITests/SamplesApp.UITests.csproj (hunk 1 above)
Comment (carldebilly) [reply src/SamplesApp/SamplesApp.UITests/SamplesApp.UITests.csproj:12]: It is.

File: src/Uno.UI/UI/Xaml/Controls/Flyout/FlyoutBase.cs [hunk 11]
```diff
@@ -20,7 +21,7 @@
 
 namespace Windows.UI.Xaml.Controls.Primitives
 {
-	public partial class FlyoutBase : DependencyObject
+	public abstract partial class FlyoutBase : DependencyObject
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Controls/Flyout/FlyoutBase.cs:24]: This should be reverted, it's not `abstract` on UWP

File: src/Uno.UI/UI/Xaml/Controls/Primitives/PickerFlyoutBase.cs [hunk 12]
```diff
@@ -4,12 +4,21 @@
 
 namespace Windows.UI.Xaml.Controls.Primitives
 {
-	public partial class PickerFlyoutBase : FlyoutBase
+	public abstract partial class PickerFlyoutBase : FlyoutBase
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Controls/Primitives/PickerFlyoutBase.cs:7]: Same here, this must not be abstract

File: src/Uno.UI/UI/Xaml/Controls/Primitives/Selector.cs [hunk 13]
```diff
@@ -480,7 +480,8 @@ protected override void OnItemsChanged(object e)
 		{
 			if (e is IVectorChangedEventArgs iVCE)
 			{
-				if (iVCE.CollectionChange == CollectionChange.ItemChanged || iVCE.CollectionChange == CollectionChange.ItemInserted)
+				if (iVCE.CollectionChange == CollectionChange.ItemChanged
+				    || (iVCE.CollectionChange == CollectionChange.ItemInserted && iVCE.Index < Items.Count))
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Controls/Primitives/Selector.cs:484]: Why did you need this - how can OnItemsChanged() be called with an inserted item that's out of range?

File: src/Uno.UI/UI/Xaml/Controls/TimePicker/TimePickerFlyout.cs [hunk 14]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using Windows.UI.Xaml.Controls.Primitives;
+
+namespace Windows.UI.Xaml.Controls
+{
+	partial class TimePickerFlyout : PickerFlyoutBase
+	{
+#if !__IOS__
+		protected override Control CreatePresenter() => throw new NotImplementedException();
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Controls/TimePicker/TimePickerFlyout.cs:9]: these will be implemented eventually? Or they're simply not supported?

File: src/Uno.UI/UI/Xaml/DependencyObjectHelper.cs [hunk 15]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Windows.UI.Xaml
+{
+	internal class DependencyObjectHelper
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/DependencyObjectHelper.cs:7]: Remove?

File: src/Uno.UWP/System/DispatcherQueue.Android.cs [hunk 16]
```diff
@@ -0,0 +1,22 @@
+﻿using Windows.UI.Core;
+
+namespace Windows.System
+{
+	partial class DispatcherQueue
+	{
+		bool TryEnqueueNative(DispatcherQueuePriority priority, DispatcherQueueHandler callback)
```
Comment (davidjohnoliver) [inline src/Uno.UWP/System/DispatcherQueue.Android.cs:7]: Why do you need this

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/DatePicker/DatePickerFlyout_Automated.xaml.cs [hunk 17]
```diff
@@ -12,7 +12,8 @@ public DatePickerFlyout_Automated()
 			InitializeComponent();
 
 			TestDatePickerFlyout.Date = new DateTimeOffset(new DateTime(2019, 3, 12));
-			TestDatePickerFlyout.DatePicked += (s, e) => selectedDate.Text = TestDatePickerFlyout.Date.ToString();
+
+			//TestDatePickerFlyout.DatePicked += (s, e) => selectedDate.Text = TestDatePickerFlyout.Date.ToString();
```
[the remark this answers is not in the document]
Comment (carldebilly) [reply src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/DatePicker/DatePickerFlyout_Automated.xaml.cs:16]: Yes

File: src/Uno.UI/UI/Xaml/Controls/CommandBar/CommandBarFlyout.cs (hunk 5 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/CommandBar/CommandBarFlyout.cs:7]: That's not something public.

Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/Controls/CommandBar/CommandBarFlyout.cs:7]: should this be a `NotSupportedException`?

File: src/Uno.UI/UI/Xaml/Controls/DatePicker/DatePicker.Flyout.cs [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+	partial class DatePicker
+	{
+
+
+#if __IOS__ || __ANDROID__
+		public static DependencyProperty UseNativeStyleProperty { get; }= DependencyProperty.Register(
+			"UseNativeStyle",
+			typeof(bool),
+			typeof(DatePicker),
+			new PropertyMetadata(!FeatureConfiguration.Style.UseUWPDefaultStyles));
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Controls/DatePicker/DatePicker.Flyout.cs:15]: Use `!Style.ShouldUseUWPDefaultStyle(typeof(DatePicker))` instead - this will also take `FeatureConfiguration.Style.UseUWPDefaultStylesOverride` into account

Even then, this approach sets up a non-obvious race between modifying the FeatureFlag and calling the `DatePicker` static constructor...

File: src/Uno.UI/UI/Xaml/Controls/ScrollContentPresenter/ScrollContentPresenter.cs [hunk 19]
```diff
@@ -82,12 +82,47 @@ bool ILayoutConstraints.IsHeightConstrained(View requester)
 
 		public double ExtentHeight
 		{
-			get => Content is FrameworkElement fe ? fe.DesiredSize.Height : 0;
+			get
+			{
+				if (Content is FrameworkElement fe)
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Controls/ScrollContentPresenter/ScrollContentPresenter.cs:87]: Interesting!
Did you add runtime tests addressing this behavior?

File: src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.Uno.cs [hunk 20]
```diff
@@ -81,7 +81,7 @@ public static void SetShouldFallBackToNativeScrollBars(Windows.UI.Xaml.Controls.
 			scrollViewer.SetValue(ShouldFallBackToNativeScrollBarsProperty, value);
 		}
 
-		public static readonly DependencyProperty ShouldFallBackToNativeScrollBarsProperty =
+		public static DependencyProperty ShouldFallBackToNativeScrollBarsProperty =
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.Uno.cs:84]: why?

File: src/Uno.UI/UI/Xaml/Controls/DatePicker/DatePicker.xaml.old [hunk 21]
```diff
@@ -0,0 +1,378 @@
+﻿<!-- Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT License. See LICENSE in the project root for license information. -->
```
[the remark this answers is not in the document]
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/DatePicker/DatePicker.xaml.old:1]: Nope...  You're right, it should be removed.

File: src/Uno.UI/UI/Xaml/Controls/Flyout/FlyoutBase.cs (hunk 11 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/Flyout/FlyoutBase.cs:24]: The constructor is protected, so it's the same...

File: src/Uno.UI/UI/Xaml/Controls/Primitives/PickerFlyoutBase.cs (hunk 12 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/Primitives/PickerFlyoutBase.cs:7]: The constructor is protected, so it's the same...

File: src/Uno.UI/UI/Xaml/Controls/Primitives/Selector.cs (hunk 13 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/Primitives/Selector.cs:484]: It was happening during my tests: sometimes the index was out of range on the items in the event args and this was crashing the app.

File: src/Uno.UI/UI/Xaml/Controls/TimePicker/TimePickerFlyout.cs (hunk 14 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/TimePicker/TimePickerFlyout.cs:9]: TimePicker not supported on other platforms right now.  Will be udpated when we'll import TimePicker from WinUI.

File: src/Uno.UWP/System/DispatcherQueue.Android.cs (hunk 16 above)
Comment (carldebilly) [reply src/Uno.UWP/System/DispatcherQueue.Android.cs:7]: DispatcherQueue is the new way to go in WinUI. All the code is using it internally. The current implementation is only a facade over the CoreDispatcher...  it sure needs to be updated in the future.

File: src/Uno.UI.RuntimeTests/MUX/Helpers/InputHelper.cs [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+		{
+			Point elementCenter = await GetElementPosition(element, fractionOfWidth, fractionOfHeight);
+
+			TapOnPoint(elementCenter);
+		}
+
+		internal static void TapOnPoint(Point point)
+		{
+			// TODO
+			throw new NotImplementedException("TapOnPoint not implemented yet");
```
[the remark this answers is not in the document]
Comment (carldebilly) [reply src/Uno.UI.RuntimeTests/MUX/Helpers/InputHelper.cs:27]: You're welcome to implement it :-)

File: src/Uno.UI.Runtime.Skia.Tizen/Tizen/TizenHost.cs [hunk 23]
```diff
@@ -59,7 +60,16 @@ public TizenHost(Func<WinUI.Application> appBuilder, string[] args = null)
...
 				.Skip(1)
 				.ToArray();
-			
+
+			bool EnqueueNative(DispatcherQueuePriority priority, DispatcherQueueHandler callback)
+			{
+				EcoreMainloop.PostAndWakeUp(() => callback());
+
+				return true;
+			}
```
[the remark this answers is not in the document]
Comment (carldebilly) [reply src/Uno.UI.Runtime.Skia.Tizen/Tizen/TizenHost.cs:75]: To make the DispatcherQueue working.

File: src/Uno.UI.Runtime.Skia.Gtk/GTK/GtkHost.cs [hunk 24]
```diff
@@ -52,32 +53,43 @@ public void Run()
...
-					   }
-					   return false;
-				   });
-			   };
+			bool EnqueueNative(DispatcherQueuePriority priority, DispatcherQueueHandler callback)
+			{
+				Dispatch(() => callback());
+
+				return true;
+			}
```
[the remark this answers is not in the document]
Comment (carldebilly) [reply src/Uno.UI.Runtime.Skia.Gtk/GTK/GtkHost.cs:61]: To make the DispatcherQueue working.

File: src/Uno.UI/UI/Xaml/Controls/ScrollContentPresenter/ScrollContentPresenter.cs (hunk 19 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/ScrollContentPresenter/ScrollContentPresenter.cs:87]: I won't have time to do it before the merge of this PR.  But sure, I'll add an issue for that.

File: src/Uno.UI/UI/Xaml/Controls/DatePicker/DatePicker.Flyout.cs (hunk 18 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/DatePicker/DatePicker.Flyout.cs:15]: I makes no sense to change it globally during the execution of an app.  And if you really want to do that, you can using a `<Style />`.

Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/DatePicker/DatePicker.Flyout.cs:15]: Fixed everything.  In my next push.

File: src/Uno.UI/UI/Xaml/DependencyObjectHelper.cs (hunk 15 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/DependencyObjectHelper.cs:7]: Yep ;-)

File: src/Uno.UI.RuntimeTests/MUX/Helpers/ControlHelper.cs (hunk 3 above)
Comment (carldebilly) [reply src/Uno.UI.RuntimeTests/MUX/Helpers/ControlHelper.cs:82]: It's code from Windows we may want to activate in the future...

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/DatePicker/DatePickerSample.xaml (hunk 2 above)
Comment (carldebilly) [reply src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/DatePicker/DatePickerSample.xaml:25]: @MartinZikmund I don't understand this one. It's bound to the VM...  it was just not working in the past.

File: src/SamplesApp/SamplesApp.UITests/SamplesApp.UITests.csproj (hunk 1 above)
Comment (carldebilly) [reply src/SamplesApp/SamplesApp.UITests/SamplesApp.UITests.csproj]: @davidjohnoliver it was an update when it was done ;-)

File: src/Uno.UI/UI/Xaml/Controls/CommandBar/CommandBarFlyout.cs (hunk 5 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/CommandBar/CommandBarFlyout.cs:7]: It's not implemented...  the difference between NotImplemented and NotSupported could be subtile:
* Not Implemented: not done yet
* Not Supported: a limitation is preventing its implementation

File: src/Uno.UI/UI/Xaml/Controls/Primitives/PickerFlyoutBase.cs (hunk 12 above)
Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/Controls/Primitives/PickerFlyoutBase.cs:7]: It's generally not a breaking change if no public methods can be called.

File: src/Uno.UI/UI/Xaml/Controls/ScrollContentPresenter/ScrollContentPresenter.cs (hunk 19 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/ScrollContentPresenter/ScrollContentPresenter.cs:87]: https://github.com/unoplatform/uno/issues/5320
